From 8b0be5922f6d53cc7864c042b6cf1088e4b6034b Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Mon, 6 May 2019 19:06:46 +0200
Subject: [PATCH 267/345] MIPS: Move VMALLOC_START to avoid OCTEON III
 Core-31034

Put VMALLOC_START at the midpoint of the MAP_BASE..VMALLOC_END range.

Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/include/asm/pgtable-64.h | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/arch/mips/include/asm/pgtable-64.h b/arch/mips/include/asm/pgtable-64.h
index 0dba7055cd00..e4b68f4896d9 100644
--- a/arch/mips/include/asm/pgtable-64.h
+++ b/arch/mips/include/asm/pgtable-64.h
@@ -145,17 +145,18 @@
 
 /*
  * TLB refill handlers also map the vmalloc area into xuseg.  Avoid
- * the first couple of pages so NULL pointer dereferences will still
- * reliably trap.
+ * the first half of the MAP_BASE area so NULL pointer dereferences
+ * will still reliably trap, and to avoid OCTEON III errata.
  */
-#define VMALLOC_START		(MAP_BASE + (2 * PAGE_SIZE))
 #define VMALLOC_END	\
 	(MAP_BASE + \
 	 min(PTRS_PER_PGD * PTRS_PER_PUD * PTRS_PER_PMD * PTRS_PER_PTE * PAGE_SIZE, \
 	     (1UL << cpu_vmbits)) - (1UL << 32))
 
-#if defined(CONFIG_MODULES) && defined(KBUILD_64BIT_SYM32) && \
-	VMALLOC_START != CKSSEG
+#define VMALLOC_START		((MAP_BASE / 2 + VMALLOC_END / 2) & PAGE_MASK)
+
+#if defined(CONFIG_MODULES) && defined(KBUILD_64BIT_SYM32)
+
 /* Load modules into 32bit-compatible segment. */
 #ifdef CONFIG_MAPPED_KERNEL
 extern unsigned long kernel_image_end;
-- 
2.25.1

